components/pthread/port/linux/pthread.c

```diff
@@ -24,20 +41,62 @@ esp_pthread_cfg_t esp_pthread_get_default_config(void)
         .prio = CONFIG_PTHREAD_TASK_PRIO_DEFAULT,
         .inherit_cfg = false,
         .thread_name = NULL,
-        .pin_to_core = 0,
-        .stack_alloc_caps = 0,
+        .pin_to_core = get_default_pthread_core(),
+        .stack_alloc_caps = MALLOC_CAP_INTERNAL | MALLOC_CAP_8BIT,
     };
 
     return cfg;
 }
 
 esp_err_t esp_pthread_set_cfg(const esp_pthread_cfg_t *cfg)
 {
+    if (cfg->stack_size < PTHREAD_STACK_MIN) {
+        return ESP_ERR_INVALID_ARG;
+    }
+
+    // 0 is treated as default value, hence change caps to MALLOC_CAP_8BIT | MALLOC_CAP_INTERNAL in that case
+    int heap_caps;
+    if (cfg->stack_alloc_caps == 0) {
+        heap_caps = MALLOC_CAP_8BIT | MALLOC_CAP_INTERNAL;
+    } else {
+        // Check that memory is 8-bit capable
+        if (!(cfg->stack_alloc_caps & MALLOC_CAP_8BIT)) {
+            return ESP_ERR_INVALID_ARG;
+        }
+
+        heap_caps = cfg->stack_alloc_caps;
+    }
+
+    /* If a value is already set, update that value */
+    esp_pthread_cfg_t *p = pthread_getspecific(s_pthread_cfg_key);
+    if (!p) {
+        p = malloc(sizeof(esp_pthread_cfg_t));
+        if (!p) {
+            return ESP_ERR_NO_MEM;
+        }
+    }
+    *p = *cfg;
+    p->stack_alloc_caps = heap_caps;
+    pthread_setspecific(s_pthread_cfg_key, p);
```

Comment:
Really minor thing: I'd suggest to add an assert here to make sure `pthread_setspecific()` succeeds, even though the chance is probably very little:

```suggestion
    int __attribute((unused)) res = pthread_setspecific(s_pthread_cfg_key, p);
    assert(res == 0);
```

You may need to `#include assert.h`.

Reply:
Ok I've tried it but it is not so straight forward, as soon as linux doesnt call esp_pthread_init, cfg_key is never created and setspecific assertion fails. I'll take a look on how to resolve that.

Reply:
@0xjakob Actually I'm thinking about all functions marked as `ESP_SYSTEM_INIT_FN`. Does it make sense to create a ticket to implement a way to run such functions in linux?

Reply:
> I've tried it but it is not so straight forward

Oh, sorry, I didn't think about that. I think you could try to use `__attribute__((constructor))` to create a constructor function to initialize the key. `__attribute__((constructor))` is documented in the [GCC function attributes doc page]( https://gcc.gnu.org/onlinedocs/gcc-13.2.0/gcc/Common-Function-Attributes.html#Common-Function-Attributes)

`ESP_SYSTEM_INIT_FN` won't work on Linux. It's not clear yet, either, if we will implement it on Linux or not. We have a ticket for exactly this in our own ticket system already. Unless you have a concrete reason to implement it, there is no need to create a ticket here.

Reply:
Ok I'll add constructor. The reason I can think to add `ESP_SYSTEM_INIT_FN` is to have more similar firmware behavior in linux than in xtensa. Maybe it could be as simple as creating a `#define ESP_SYSTEM_INIT_FN(foo) __attribute__((constructor)) foo`.

Reply:
@0xjakob this is ready for review